torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -0,0 +1,211 @@
+    }
+  } else {
+    st = PromoteScalarTypes(typesFromScalars);
+  }
+
+  return st;
+}
+
+static void UpdateScalarTypeForInputs(Node* n, const c10::ScalarType& scalar_type) {
+  const int64_t onnx_type = ScalarTypeToONNXType(scalar_type);
+  if (onnx_type < 0) {
+    std::cerr << "Warning: ONNX Scalar Type Analysis - Scalar type: "
+              << c10::toString(scalar_type)
+              << " of input tensor in operator: " << n->kind().toDisplayString()
+              << " not supported in ONNX. " << std::endl;
+    return;
+  }
+
+  for (auto input : n->inputs()) {
+    auto input_tensor_type = ProfiledTensorType::create(input->type());
+    auto input_scalar_type = input_tensor_type->scalarType();
+
+    if ((input->node()->kind() == onnx::Constant) ||
+        (input_scalar_type && (*input_scalar_type != scalar_type))) {
+      if (input->node()->kind() == onnx::Constant) {
+        // Fix up the scalar directly instead of inserting a cast operator.
+        // NOTE: Keep only the else branch once constant_folding is enabled by
+        // default.
+        at::Tensor val = input->node()->t(attr::value);
+        at::Tensor new_val = val.to(scalar_type);
+        Node* const_node = n->owningGraph()->create(onnx::Constant);
+        const_node->t_(attr::value, new_val);
+        const_node->insertBefore(n);
+        const_node->output()->setType(ProfiledTensorType::create(new_val));
+        n->replaceInputWith(input, const_node->output());
+      } else {
+        Node* cast_node = n->owningGraph()->create(onnx::Cast);
+        cast_node->addInput(input);
+        cast_node->i_(attr::to, onnx_type);
+        cast_node->insertBefore(n);
+        cast_node->output()->setType(CreateProfiledTensorTypeWithScalarType(
+            input_tensor_type, scalar_type));
+        n->replaceInputWith(input, cast_node->output());
+      }
+    }
+  }
+}
+
+static void UpdateScalarTypeForOutput(Node* n, const c10::ScalarType& scalar_type) {
+  auto output_tensor_type = ProfiledTensorType::create(n->output()->type());
+  n->output()->setType(
+      CreateProfiledTensorTypeWithScalarType(output_tensor_type, scalar_type));
+}
+
+static void ImplicitCastForONNX(Block* block) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
+    for (auto sub : it->blocks()) {
+      ImplicitCastForONNX(sub);
+    }
+    WithInsertPoint guard(*it);
+    auto* subgraph = it->owningGraph();
+
+    if (IsImplicitCastSupported(it->kind())) {
+      auto expected_scalar_type = InferExpectedScalarType(*it);
+      if (expected_scalar_type) {
+        UpdateScalarTypeForInputs(*it, *expected_scalar_type);
+        if (!IsComparisonOp(it->kind())) {
+          UpdateScalarTypeForOutput(*it, *expected_scalar_type);
+        }
+      }
+    }
+  }
+  EliminateDeadCode(block, true, DCESideEffectPolicy::ALLOW_DELETING_NODES_WITH_SIDE_EFFECTS);
+}
+
+// This pass tries to resolve scalar type mismatch issues between input tensors
```

Comment:
Is it correct to say that we are adding this as an ONNX post-pass so that if there are any potential mismatches that come up in ONNX graph creation through symbolic entry points, this pass will address those as well?

Reply:
Yes, that is one of the motivations for adding this as post onnx pass.